mixer/pkg/api/grpcServer.go

```diff
@@ -212,6 +212,7 @@ func (s *grpcServer) check(ctx context.Context, req *mixerpb.CheckRequest,
 			} else {
 				if !status.IsOK(qr.Status) {
 					lg.Debugf("Quota denied: %v", qr.Status)
+					resp.Precondition.Status = qr.Status
```

Comment:
@geeknoid, @douglas-reid : Does this looks like the right way to pass error from quota other than resource exhausted to client? I don't see how we are passing err to envoy mixer client... 
I think having a better function here, which looks at all errors from all quotas and if there is a resource exhausted error, we pass-in that as compared to last one wins kinda logic is better..
Please let me know what you guys think...

Reply:
Not forwarding the status here was intentional. There is a split between precondition results and quota results, this mingles the two together.

The issue is really that there's a mismatch between the Mixer->Adapter protocol vs. the Envoy->Mixer protocol. In the Mixer->Adapter protocol there are three different outputs:

- An error value
- A bunch of tokens
- A status value

But in the Envoy->Mixer protocol only the first two are covered. I had meant to remove the QuotaResult.Status field completely, but I guess I never got around to it.

So basically, I think the adapter should either return an error value indicating it failed the attempt to allocate quota, or it should return an OK status and the allocated quota amount, which can legitimately be 0.

Reply:
@geeknoid : where is error returned used in Envoy->Mixer protocol? I don't see it getting used:(

Reply:
I'm not familiar with the mixer client code. It's quite possible that the error checking logic was never put in place :-(

Reply:
Adding @qiwzhang , @kyessenov who are more familiar with the client code... 
@qiwzhang , @kyessenov : is error returned by Mixer->Adapter used in client code?